docs/proposal/provider-metrics.md

```diff
@@ -0,0 +1,80 @@
+# Exporting Provider-Specific Metrics in External-DNS
+*(December 2023)*
+
+## Purpose
+- The purpose of this proposal is to introduce the functionality to export provider-specific metrics in the external-dns project.
+- This enhancement will enable users to gain insights into provider-specific behavior, leading to better management and troubleshooting of DNS services.
+
+## Motivating Usecases
+- **Rate Limit Monitoring** - Many providers implement API rate-limits. Per-provider metrics would enable users to detect and respond when they are approaching these limits:
+   - Cloudflare: [#2135](https://github.com/kubernetes-sigs/external-dns/issues/2135)
+   - AWS: [#1293](https://github.com/kubernetes-sigs/external-dns/issues/1293)
+   - DigitalOcean: [#1429](https://github.com/kubernetes-sigs/external-dns/issues/1429)
+
+- **Performance/Latency Monitoring** - Changes to the volume/content of provider API requests may impact performance of external-dns overall. Per-provider metrics can help identify and debug these regressions.
+  - AWS: [#869](https://github.com/kubernetes-sigs/external-dns/issues/869)
+
+- **Debugging** - Per-provider metrics allow for specific failure conditions to be surfaced to users via metrics as opposed to logs. This could expidite investigations.
+
+_This list is non-exhaustive_
+
+## High-Level Design
+
+### Metric Format
+Provider-specific metrics will be exported with the format  `external_dns_provider_{metric_name}`. They
+should have a `provider` label that indicates the provider the metric is associated with. 
+
+Certain types of metrics (such as request counts, latency) are expected to conform to a common specification to make the experience consistent across providers:
+
+- `external_dns_provider_request_count`
+  - Counts the number of requests issued.
+  - Type: counter
+  - Labels:
+     - `method` - The method being invoked (eg. `ListTagsForResourceWithContext`).
+     - `error` - Whether the invocation of this method resulted in an error.
+     - `error_code` - Error code returned from the method. 
```

Comment:
do we always get an error_code?
I think the SDKs do not return an error code but an error.

Reply:
I will clarify in the doc- Not all SDKs will return all labels, but if the SDK returns that specific label, it should be exported.